### PR TITLE
Fixes for "show payment on confirmation page" for events

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
See https://lab.civicrm.org/extensions/stripe/-/issues/445

Before
----------------------------------------
Payments are incomplete if pay later enabled

After
----------------------------------------
Works properly with Stripe. Stripe + Pay later. Stripe + other processor.

Technical Details
----------------------------------------

Comments
----------------------------------------

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@mattwire I'm not looking at the event forms at the moment but I am looking at the contribution pages - is this something that would apply to them too? Can you tell me what you are working on?

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
> @mattwire I'm not looking at the event forms at the moment but I am looking at the contribution pages - is this something that would apply to them too? Can you tell me what you are working on?

This only applies to the "show payment on confirmation page" functionality which has a couple of edge-cases that are broken.

**Comment by mattwire** [comment]
test this please

File: CRM/Event/Form/Registration/Confirm.php [hunk 1]
```diff
@@ -456,6 +457,15 @@ public function postProcess() {
     $participantCount = [];
     $totalTaxAmount = 0;
 
+    if ($this->isShowPaymentOnConfirm()) {
+      // We assign this here so we don't assign it multiple times in the foreach below
+      $submitValues = $this->getSubmitValues();
```
Comment (eileenmcnaughton) [inline CRM/Event/Form/Registration/Confirm.php:462]: @mattwire `getSubmittedValues()` is the safe method  -or `$this->_paymentProcessorIDs = [$this->getSubmittedValues('payment_processor_id')];` - it handles money formatting & checks the field is present - the other basically just gets `_POST`

Comment (eileenmcnaughton) [reply CRM/Event/Form/Registration/Confirm.php:464]: How is it getting into submitValues but not submittedValues - that would happen if the field were not added to the form but somehow posting anyway? Or is it being added to the `Confirm` form but this function is too 'Main-biased' https://github.com/civicrm/civicrm-core/blob/8dd4dec4e0ee44d53ebbd1681a21d16431061e6d/CRM/Contribute/Form/ContributionBase.php#L1402

Ideally we would add a function `getPaymentProcessorID()` as that is the pattern we are now adding - if that is there then including `CRM_Financial_Form_PaymentProcessorFormTrait` would give us `getPaymentProcessorValue()` too

Comment (eileenmcnaughton) [reply CRM/Event/Form/Registration/Confirm.php:464]: @mattwire where are the fields added to the form?

Comment (eileenmcnaughton) [reply CRM/Event/Form/Registration/Confirm.php]: @mattwire ah ok - you are not calling that `buildQuickForm()` when `isSubmitted()` is true - which would prevent the fields being added to the quickForm - hence `$this->getSubmittedValue('payment_processor_id')` doesn't find the field on the form & so filters it out

Comment (eileenmcnaughton) [reply CRM/Event/Form/Registration/Confirm.php]: @mattwire it might be worth adding a variant of https://github.com/civicrm/civicrm-core/blob/12cd7925fd9f752173d3d39781fe29e25814e298/CRM/Contribute/Form/ContributionBase.php#L1353-L1375 to the Event form flow now - because if it gets added later other people (like me) might not recall that values can now be submitted on the Confirm page & that would cause a regression that affects you.

It could be much like that one but just also merge in Confirm fields

Comment (mattwire) [reply CRM/Event/Form/Registration/Confirm.php]: I didn't add a `getSubmittedValue` function. I'm not sure it makes sense here as there are too many forms involved that can submit values (eg. "Register", "additionalparticipant", "Confirm") and maybe backend forms?

But I have resolved the use of the generic `getSubmittedValue()` so should be ok now?

**Comment by eileenmcnaughton** [comment]
@mattwire can you confirm this addresses the issue in https://github.com/civicrm/civicrm-core/pull/30313

**Comment by mattwire** [comment]
> @mattwire can you confirm this addresses the issue in #30313

It's related but a slightly different issue.

This PR fixes various combinations of pay later and multiple payment processors not working correctly when payment is on the confirmation page.

**Comment by mattwire** [comment]
Thanks @eileenmcnaughton
